Now we will look how to create and perform operations on dataframes and understand what it does or looks like.

In [1]:
import pandas as pd
import numpy as np
from numpy.random import randn
my_data = randn(4,3) # 4 rows, 3 columns
my_rows = ['row1', 'row2', 'row3', 'row4']
my_cols = ['monday', 'tuesday', 'wednesday']

Now we will actually create the dataframe and parameters go like data,rows,columns

In [2]:
my_df = pd.DataFrame(data=my_data, index=my_rows, columns=my_cols)
print(my_df)

        monday   tuesday  wednesday
row1 -0.216712  1.627002  -0.183485
row2  0.159469 -1.005483  -0.858452
row3  1.140222 -1.786657   0.821717
row4  0.237223 -2.221154  -1.103862


Now we will import a given dataset that is already a csv file and convert it directly to a dataframe.
pd.read_csv will do this whole operation for us and save us the data pulling and pushing step. I will use the already existing iris data that I used in the neural network notebook in the same repo.

In [3]:
my_df2 = pd.read_csv('Data/iris.csv')
print(my_df2)

     sepal_length  sepal_width  petal_length  petal_width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]


We can now pull out rows using the index numbers in this dataframe using loc()

In [4]:
print(my_df2.loc[50])

sepal_length                7.0
sepal_width                 3.2
petal_length                4.7
petal_width                 1.4
class           Iris-versicolor
Name: 50, dtype: object


Now we will look at accessing data from this entire dataframe:
1. First 5 rows
2. First n rows
3. Last 5 rows
4. Last n rows
5. Get context/info

In [5]:
my_df2.head() 

sepal_length  sepal_width  petal_length  petal_width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [6]:
my_df2.head(10)

sepal_length  sepal_width  petal_length  petal_width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa
5           5.4          3.9           1.7          0.4  Iris-setosa
6           4.6          3.4           1.4          0.3  Iris-setosa
7           5.0          3.4           1.5          0.2  Iris-setosa
8           4.4          2.9           1.4          0.2  Iris-setosa
9           4.9          3.1           1.5          0.1  Iris-setosa

In [7]:
my_df2.tail()

sepal_length  sepal_width  petal_length  petal_width           class
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

In [8]:
my_df2.tail(3)

sepal_length  sepal_width  petal_length  petal_width           class
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

In [9]:
my_df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   class         150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


Now we will look at these below features:
1. Number of dimensions in dataframe
2. Datatypes of columns alone (Not like info)
3. Get basic stats like avg, 25%, 50% and all that kind of information ---> Do it for specific columns


In [10]:
my_df2.ndim

2

In [11]:
my_df2.dtypes

sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
class            object
dtype: object

In [12]:
my_df2['sepal_length'].describe() #for one column

count    150.000000
mean       5.843333
std        0.828066
min        4.300000
25%        5.100000
50%        5.800000
75%        6.400000
max        7.900000
Name: sepal_length, dtype: float64

Now we will look at counting operations on these dataframes.
1. Count number of something in descending order (Default in value_counts()). Just pass ascending=True in the value_counts function to reverse this order. 

In [13]:
my_df2['class'].value_counts()

class
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
Name: count, dtype: int64

2. By default if there is a none value (NaN), value_counts will not return it. To include it use dropna=False as parameter for value_counts()

We can also get the percentage of any single column by setting Normalize=True.

In [14]:
my_df2['sepal_length'].value_counts(normalize=True)

sepal_length
5.0    0.066667
6.3    0.060000
5.1    0.060000
6.7    0.053333
5.7    0.053333
6.4    0.046667
5.5    0.046667
5.8    0.046667
4.9    0.040000
6.0    0.040000
5.4    0.040000
5.6    0.040000
6.1    0.040000
6.5    0.033333
4.8    0.033333
7.7    0.026667
6.9    0.026667
4.6    0.026667
5.2    0.026667
6.2    0.026667
4.4    0.020000
7.2    0.020000
5.9    0.020000
6.8    0.020000
4.7    0.013333
6.6    0.013333
4.3    0.006667
7.0    0.006667
5.3    0.006667
4.5    0.006667
7.1    0.006667
7.3    0.006667
7.6    0.006667
7.4    0.006667
7.9    0.006667
Name: proportion, dtype: float64

Single item counts can also be called from a specific column

In [15]:
my_df2['class'].value_counts()['Iris-setosa']

np.int64(50)

We can also add new columns to our dataframes using multiple methods
1. List: This method is hard for big datasets like this one because I will have to add a list of 150 elements
2. Default: Can be used for big data sets. This issue is this can only be used if all of the elements in new column are same so mostly used for obvious yes or no questions

In [16]:
my_df2['alive'] = ['yes'] * len(my_df2)
print(my_df2)

     sepal_length  sepal_width  petal_length  petal_width           class  \
0             5.1          3.5           1.4          0.2     Iris-setosa   
1             4.9          3.0           1.4          0.2     Iris-setosa   
2             4.7          3.2           1.3          0.2     Iris-setosa   
3             4.6          3.1           1.5          0.2     Iris-setosa   
4             5.0          3.6           1.4          0.2     Iris-setosa   
..            ...          ...           ...          ...             ...   
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

    alive  
0     yes  
1     yes  
2     yes  
3     yes  
4     yes  
.. 

We can also add a NaN or null value to a whole column

In [17]:
my_df2['found everywhere?']= [np.nan] * len(my_df2)
print(my_df2)

     sepal_length  sepal_width  petal_length  petal_width           class  \
0             5.1          3.5           1.4          0.2     Iris-setosa   
1             4.9          3.0           1.4          0.2     Iris-setosa   
2             4.7          3.2           1.3          0.2     Iris-setosa   
3             4.6          3.1           1.5          0.2     Iris-setosa   
4             5.0          3.6           1.4          0.2     Iris-setosa   
..            ...          ...           ...          ...             ...   
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

    alive  found everywhere?  
0     yes                NaN  
1     yes    

The insert function will allow us to put a column in a certain position of dataframe not only at the end

In [18]:
my_df2.insert(2, 'bought?', ['no'] * len(my_df2))
print(my_df2)

     sepal_length  sepal_width bought?  petal_length  petal_width  \
0             5.1          3.5      no           1.4          0.2   
1             4.9          3.0      no           1.4          0.2   
2             4.7          3.2      no           1.3          0.2   
3             4.6          3.1      no           1.5          0.2   
4             5.0          3.6      no           1.4          0.2   
..            ...          ...     ...           ...          ...   
145           6.7          3.0      no           5.2          2.3   
146           6.3          2.5      no           5.0          1.9   
147           6.5          3.0      no           5.2          2.0   
148           6.2          3.4      no           5.4          2.3   
149           5.9          3.0      no           5.1          1.8   

              class alive  found everywhere?  
0       Iris-setosa   yes                NaN  
1       Iris-setosa   yes                NaN  
2       Iris-setosa   yes     

The assign function can also be used to add a whole new attribute the only thing is it will add this column and create a whole new dataframe so we should change the variable associated with the operation with which we can call on both datasets seperately

In [19]:
my_df3 = my_df2.assign(Colour=['blue'] * len(my_df2))
print(my_df3)

     sepal_length  sepal_width bought?  petal_length  petal_width  \
0             5.1          3.5      no           1.4          0.2   
1             4.9          3.0      no           1.4          0.2   
2             4.7          3.2      no           1.3          0.2   
3             4.6          3.1      no           1.5          0.2   
4             5.0          3.6      no           1.4          0.2   
..            ...          ...     ...           ...          ...   
145           6.7          3.0      no           5.2          2.3   
146           6.3          2.5      no           5.0          1.9   
147           6.5          3.0      no           5.2          2.0   
148           6.2          3.4      no           5.4          2.3   
149           5.9          3.0      no           5.1          1.8   

              class alive  found everywhere? Colour  
0       Iris-setosa   yes                NaN   blue  
1       Iris-setosa   yes                NaN   blue  
2       I

In [20]:
print(my_df2.columns)

Index(['sepal_length', 'sepal_width', 'bought?', 'petal_length', 'petal_width',
       'class', 'alive', 'found everywhere?'],
      dtype='object')


We can see that the colour attribute or column is not there in the original dataframe

We can also similarly delete columns and rows from a dataframe. Use the simple drop function.

In [21]:
my_df2.drop("alive", axis=1, inplace=False) #axis 1 means the header row
print(my_df2)

     sepal_length  sepal_width bought?  petal_length  petal_width  \
0             5.1          3.5      no           1.4          0.2   
1             4.9          3.0      no           1.4          0.2   
2             4.7          3.2      no           1.3          0.2   
3             4.6          3.1      no           1.5          0.2   
4             5.0          3.6      no           1.4          0.2   
..            ...          ...     ...           ...          ...   
145           6.7          3.0      no           5.2          2.3   
146           6.3          2.5      no           5.0          1.9   
147           6.5          3.0      no           5.2          2.0   
148           6.2          3.4      no           5.4          2.3   
149           5.9          3.0      no           5.1          1.8   

              class alive  found everywhere?  
0       Iris-setosa   yes                NaN  
1       Iris-setosa   yes                NaN  
2       Iris-setosa   yes     

But this is not permanent deletion we can see in the next cell.

In [22]:
print(my_df2.columns)

Index(['sepal_length', 'sepal_width', 'bought?', 'petal_length', 'petal_width',
       'class', 'alive', 'found everywhere?'],
      dtype='object')


To permanently delete we use inplace as True pands does this to avoid mistakes.

In [23]:
my_df2.drop('alive', axis=1, inplace=True)
print(my_df2)

     sepal_length  sepal_width bought?  petal_length  petal_width  \
0             5.1          3.5      no           1.4          0.2   
1             4.9          3.0      no           1.4          0.2   
2             4.7          3.2      no           1.3          0.2   
3             4.6          3.1      no           1.5          0.2   
4             5.0          3.6      no           1.4          0.2   
..            ...          ...     ...           ...          ...   
145           6.7          3.0      no           5.2          2.3   
146           6.3          2.5      no           5.0          1.9   
147           6.5          3.0      no           5.2          2.0   
148           6.2          3.4      no           5.4          2.3   
149           5.9          3.0      no           5.1          1.8   

              class  found everywhere?  
0       Iris-setosa                NaN  
1       Iris-setosa                NaN  
2       Iris-setosa                NaN  
3      

Now it's permanently gone. We can do the sae operation to rows by passing it's index as a parameter in drop

In [24]:
my_df2.drop(2)
print(my_df2)

     sepal_length  sepal_width bought?  petal_length  petal_width  \
0             5.1          3.5      no           1.4          0.2   
1             4.9          3.0      no           1.4          0.2   
2             4.7          3.2      no           1.3          0.2   
3             4.6          3.1      no           1.5          0.2   
4             5.0          3.6      no           1.4          0.2   
..            ...          ...     ...           ...          ...   
145           6.7          3.0      no           5.2          2.3   
146           6.3          2.5      no           5.0          1.9   
147           6.5          3.0      no           5.2          2.0   
148           6.2          3.4      no           5.4          2.3   
149           5.9          3.0      no           5.1          1.8   

              class  found everywhere?  
0       Iris-setosa                NaN  
1       Iris-setosa                NaN  
2       Iris-setosa                NaN  
3      

This isn't permanent deletion, we should give the inplace argument for this as well. Also axis is 0 by default in row deletion. 

In [25]:
my_df2.drop(2, inplace=True)
print(my_df2)

     sepal_length  sepal_width bought?  petal_length  petal_width  \
0             5.1          3.5      no           1.4          0.2   
1             4.9          3.0      no           1.4          0.2   
3             4.6          3.1      no           1.5          0.2   
4             5.0          3.6      no           1.4          0.2   
5             5.4          3.9      no           1.7          0.4   
..            ...          ...     ...           ...          ...   
145           6.7          3.0      no           5.2          2.3   
146           6.3          2.5      no           5.0          1.9   
147           6.5          3.0      no           5.2          2.0   
148           6.2          3.4      no           5.4          2.3   
149           5.9          3.0      no           5.1          1.8   

              class  found everywhere?  
0       Iris-setosa                NaN  
1       Iris-setosa                NaN  
3       Iris-setosa                NaN  
4      

And like that the 2nd indexed row will be dropped.

We can also grab rows, points and some subsets of data in a dataframe. We can use the index of the row for this

In [26]:
my_df2.loc[10] # If the first column hadnames and not numbers, you would use the name instead of 10

sepal_length                 5.4
sepal_width                  3.7
bought?                       no
petal_length                 1.5
petal_width                  0.2
class                Iris-setosa
found everywhere?            NaN
Name: 10, dtype: object

We use iloc to use purely index values for this 

In [27]:
my_df2.iloc[10]

sepal_length                 4.8
sepal_width                  3.4
bought?                       no
petal_length                 1.6
petal_width                  0.2
class                Iris-setosa
found everywhere?            NaN
Name: 11, dtype: object

If we want only a single feature let's say sepal_length of a specific flower let's say the row label is 4 then we can also do that

In [28]:
my_df2.loc[1, "sepal_length"]

np.float64(4.9)

If we want multiple of these we can use subsets for these. That is like 2 or 3 specific rows from which we need different different columns respectively

In [29]:
my_df2.loc[[1, 3, 67], ["sepal_length", "petal_width", "petal_length"]]

sepal_length  petal_width  petal_length
1            4.9          0.2           1.4
3            4.6          0.2           1.5
67           5.8          1.0           4.1

We can do conditional selections in pandas just like the 'WHERE' parameter in mysql to iterate and obtain specific data.

In [30]:
my_df2 == "Iris-setosa"

sepal_length  sepal_width  bought?  petal_length  petal_width  class  \
0           False        False    False         False        False   True   
1           False        False    False         False        False   True   
3           False        False    False         False        False   True   
4           False        False    False         False        False   True   
5           False        False    False         False        False   True   
..            ...          ...      ...           ...          ...    ...   
145         False        False    False         False        False  False   
146         False        False    False         False        False  False   
147         False        False    False         False        False  False   
148         False        False    False         False        False  False   
149         False        False    False         False        False  False   

     found everywhere?  
0                False  
1                False  
3                False  
4                False  
5                False  
..                 ...  
145              False  
146              False  
147              False  
148              False  
149              False  

[149 rows x 7 columns]

Unlike sql where only those rows or datapoints are given out, here it obtains the whole dataframe but assigns boolean values (T/F) to the parameter that is true

To return the actual data that is to be obtained we can simply do this.

In [31]:
my_df2[my_df2 == "Iris-setosa"]

sepal_length  sepal_width bought?  petal_length  petal_width  \
0             NaN          NaN     NaN           NaN          NaN   
1             NaN          NaN     NaN           NaN          NaN   
3             NaN          NaN     NaN           NaN          NaN   
4             NaN          NaN     NaN           NaN          NaN   
5             NaN          NaN     NaN           NaN          NaN   
..            ...          ...     ...           ...          ...   
145           NaN          NaN     NaN           NaN          NaN   
146           NaN          NaN     NaN           NaN          NaN   
147           NaN          NaN     NaN           NaN          NaN   
148           NaN          NaN     NaN           NaN          NaN   
149           NaN          NaN     NaN           NaN          NaN   

           class  found everywhere?  
0    Iris-setosa                NaN  
1    Iris-setosa                NaN  
3    Iris-setosa                NaN  
4    Iris-setosa                NaN  
5    Iris-setosa                NaN  
..           ...                ...  
145          NaN                NaN  
146          NaN                NaN  
147          NaN                NaN  
148          NaN                NaN  
149          NaN                NaN  

[149 rows x 7 columns]

All others become NULL values. If only the class column is what we need and not the whole dataframe.

In [32]:
my_df2[my_df2 == "Iris-setosa"]['class']

0      Iris-setosa
1      Iris-setosa
3      Iris-setosa
4      Iris-setosa
5      Iris-setosa
          ...     
145            NaN
146            NaN
147            NaN
148            NaN
149            NaN
Name: class, Length: 149, dtype: object

For multiple columns we can pass multiple element list in the above statement

In [33]:
my_df2[my_df2 == "Iris-setosa"][["class","sepal_length","petal_width"]]

class  sepal_length  petal_width
0    Iris-setosa           NaN          NaN
1    Iris-setosa           NaN          NaN
3    Iris-setosa           NaN          NaN
4    Iris-setosa           NaN          NaN
5    Iris-setosa           NaN          NaN
..           ...           ...          ...
145          NaN           NaN          NaN
146          NaN           NaN          NaN
147          NaN           NaN          NaN
148          NaN           NaN          NaN
149          NaN           NaN          NaN

[149 rows x 3 columns]

But all the sepal_length and petal_width don't have an element called iris-setosa it gives NaN otherwise it would give the value passed. We can also give multiple conditions in pandas using & sign

In [34]:
my_df2[(my_df2["class"] == "Iris-setosa") & (my_df2["sepal_length"] == 5.1)]

sepal_length  sepal_width bought?  petal_length  petal_width        class  \
0            5.1          3.5      no           1.4          0.2  Iris-setosa   
17           5.1          3.5      no           1.4          0.3  Iris-setosa   
19           5.1          3.8      no           1.5          0.3  Iris-setosa   
21           5.1          3.7      no           1.5          0.4  Iris-setosa   
23           5.1          3.3      no           1.7          0.5  Iris-setosa   
39           5.1          3.4      no           1.5          0.2  Iris-setosa   
44           5.1          3.8      no           1.9          0.4  Iris-setosa   
46           5.1          3.8      no           1.6          0.2  Iris-setosa   

    found everywhere?  
0                 NaN  
17                NaN  
19                NaN  
21                NaN  
23                NaN  
39                NaN  
44                NaN  
46                NaN

We can also count the number of records that satisfy the condition passed 

In [35]:
len(my_df2[(my_df2["class"] == "Iris-setosa") & (my_df2["sepal_length"] == 5.1)])

8

We can also use the 'OR' clause for either this or that using | operator

In [36]:
my_df2[(my_df2["sepal_length"] == 5.0) | (my_df2["sepal_length"] == 5.1)]

sepal_length  sepal_width bought?  petal_length  petal_width  \
0            5.1          3.5      no           1.4          0.2   
4            5.0          3.6      no           1.4          0.2   
7            5.0          3.4      no           1.5          0.2   
17           5.1          3.5      no           1.4          0.3   
19           5.1          3.8      no           1.5          0.3   
21           5.1          3.7      no           1.5          0.4   
23           5.1          3.3      no           1.7          0.5   
25           5.0          3.0      no           1.6          0.2   
26           5.0          3.4      no           1.6          0.4   
35           5.0          3.2      no           1.2          0.2   
39           5.1          3.4      no           1.5          0.2   
40           5.0          3.5      no           1.3          0.3   
43           5.0          3.5      no           1.6          0.6   
44           5.1          3.8      no           1.9          0.4   
46           5.1          3.8      no           1.6          0.2   
49           5.0          3.3      no           1.4          0.2   
60           5.0          2.0      no           3.5          1.0   
93           5.0          2.3      no           3.3          1.0   
98           5.1          2.5      no           3.0          1.1   

              class  found everywhere?  
0       Iris-setosa                NaN  
4       Iris-setosa                NaN  
7       Iris-setosa                NaN  
17      Iris-setosa                NaN  
19      Iris-setosa                NaN  
21      Iris-setosa                NaN  
23      Iris-setosa                NaN  
25      Iris-setosa                NaN  
26      Iris-setosa                NaN  
35      Iris-setosa                NaN  
39      Iris-setosa                NaN  
40      Iris-setosa                NaN  
43      Iris-setosa                NaN  
44      Iris-setosa                NaN  
46      Iris-setosa                NaN  
49      Iris-setosa                NaN  
60  Iris-versicolor                NaN  
93  Iris-versicolor                NaN  
98  Iris-versicolor                NaN

If we want only the class of flowers that satisfy the given or or and condition we can do that using the specific column feature 

In [37]:
my_df2[(my_df2["sepal_length"] == 5.0) | (my_df2["sepal_length"] == 5.1)]["class"]

0         Iris-setosa
4         Iris-setosa
7         Iris-setosa
17        Iris-setosa
19        Iris-setosa
21        Iris-setosa
23        Iris-setosa
25        Iris-setosa
26        Iris-setosa
35        Iris-setosa
39        Iris-setosa
40        Iris-setosa
43        Iris-setosa
44        Iris-setosa
46        Iris-setosa
49        Iris-setosa
60    Iris-versicolor
93    Iris-versicolor
98    Iris-versicolor
Name: class, dtype: object